array_sparse.go

```diff
@@ -512,9 +539,23 @@ func (a *sparseArrayObject) MemUsage(ctx *MemUsageContext) (memUsage uint64, err
 	// sparseArrayObject overhead
 	memUsage = SizeEmptyStruct
 
+	inc, err := a.baseObject.MemUsage(ctx)
+	memUsage += inc
+
+	if ctx.ArrayLenExceedsThreshold(len(a.items)) {
+		inc, err := a.estimateMemUsage(ctx)
+		memUsage += inc
+		if err != nil {
+			return memUsage, err
+		}
+
+		ctx.Ascend()
```

Comment:
[q] is the ctx.Ascend and Descend used as a stack to calculate the mem usage for every items for each level? How come we need to explicitly put ctx.Ascend() here? Do we need to do a defer ctx.Ascend so it pops out if there are any errors?

Reply:
You make a good point but I'd rather make this change separately so it's easy to revert.

Reply:
Created https://jira.mongodb.org/browse/BAAS-32009